monitor/events.go

```diff
@@ -3,76 +3,263 @@ package monitor
+			"startSeq":           startSeq,
 		},
 	}
 
 	sendPost(e)
 }
 
-func LogStreamEndedEvent(manifestID string, nonce uint64) {
-	e := &Event{
-		Name: "StreamEnded",
-		Properties: map[string]string{
-			"manifestID": manifestID,
-			"nonce":      fmt.Sprintf("%v", nonce),
+func LogStreamCreatedEvent(hlsStrmID string, nonce uint64) {
+	e := &event{
+		Name:  "StreamCreated",
+		Nonce: strconv.FormatUint(nonce, 10),
+		Properties: map[string]interface{}{
+			"hlsStrmID": hlsStrmID,
 		},
 	}
 
 	sendPost(e)
 }
 
-func LogStreamCreateFailed(manifestID string, nonce uint64, reason string) {
+func LogStreamStartedEvent(nonce uint64) {
+	e := &event{
+		Name:  "StreamStarted",
+		Nonce: strconv.FormatUint(nonce, 10),
+	}
+
+	sendPost(e)
+}
+
+func LogStreamEndedEvent(nonce uint64) {
+	e := &event{
+		Name:  "StreamEnded",
+		Nonce: strconv.FormatUint(nonce, 10),
+	}
+
+	sendPost(e)
+}
+
+func LogStreamCreateFailed(nonce uint64, reason string) {
 	glog.Infof("Logging StreamCreateFailed...")
-	e := &Event{
-		Name: "StreamCreateFailed",
-		Properties: map[string]string{
-			"manifestID": manifestID,
-			"reason":     reason,
-			"nonce":      fmt.Sprintf("%v", nonce),
+	e := &event{
+		Name:  "StreamCreateFailed",
+		Nonce: strconv.FormatUint(nonce, 10),
+		Properties: map[string]interface{}{
+			"reason": reason,
 		},
 	}
 
 	sendPost(e)
 }
 
-func sendPost(e *Event) {
+func LogSegmentUploadFailed(nonce, seqNo uint64, reason string) {
+	glog.Infof("Logging LogSegmentUploadFailed...")
+	e := &event{
+		Name:  "SegmentUploadFailed",
+		Nonce: strconv.FormatUint(nonce, 10),
+		Properties: map[string]interface{}{
+			"reason": reason,
+			"seqNo":  seqNo,
+		},
+	}
+
+	sendPost(e)
+}
+
+func LogSegmentEmerged(nonce, seqNo uint64) {
+	var sincePrevious time.Duration
+	now := time.Now()
+	if metrics.lastSegmentNonce == nonce {
+		sincePrevious = now.Sub(metrics.lastSegmentEmergedAt)
+	} else {
+		metrics.segmentsInFlight = 0
+		metrics.lastSegmentNonce = nonce
+		metrics.lastSeqNo = int64(seqNo) - 1
+	}
+	metrics.lastSegmentEmergedAt = now
+	e := &event{
+		Name:  "SegmentEmerged",
+		Nonce: strconv.FormatUint(nonce, 10),
+		Properties: map[string]interface{}{
+			"seqNo":         seqNo,
+			"sincePrevious": uint64(sincePrevious / time.Millisecond),
+		},
+	}
+
+	sendPost(e)
+}
+
+func SegmentUploadStart(nonce, seqNo uint64) {
+	if metrics.lastSegmentNonce == nonce {
+		metrics.segmentsInFlight++
+	}
+}
+
+func LogSegmentUploaded(nonce, seqNo uint64, uploadDur time.Duration) {
+	e := &event{
+		Name:  "SegmentUploaded",
+		Nonce: strconv.FormatUint(nonce, 10),
+		Properties: map[string]interface{}{
+			"seqNo":          seqNo,
+			"uploadDuration": uint64(uploadDur / time.Millisecond),
+		},
+	}
+
+	sendPost(e)
+}
+
+func detectSeqDif(e *event, nonce, seqNo uint64) {
+	if metrics.lastSegmentNonce == nonce {
+		seqDif := int64(seqNo) - metrics.lastSeqNo
+		metrics.lastSeqNo = int64(seqNo)
+		if seqDif != 1 {
+			e.Properties["seqNoDif"] = seqDif
```

Comment:
Likewise, this is the sort of thing that could be done on the server to reduce local bookkeeping. But we can keep this here for now and compare results.